File: java-devildyw/src/main/java/com/dyw/leetcode/algorithm/贪心/L122买卖股票的最佳时机II.java
package com.dyw.leetcode.algorithm.贪心;

/**
 * 给你一个整数数组 prices ，其中prices[i] 表示某支股票第 i 天的价格。
 * <p>
 * 在每一天，你可以决定是否购买和/或出售股票。你在任何时候最多只能持有 一股 股票。你也可以先购买，然后在 同一天 出售。
 * <p>
 * 返回 你能获得的 最大 利润。
 * <p>
 *  注意：
 * 只有一只股票！
 * 当前只有买股票或者卖股票的操作
 * 想获得利润至少要两天为一个交易单元。
 *
 * @author Devil
 * @since 2022-10-18-10:25
 */
@SuppressWarnings("all")
public class L122买卖股票的最佳时机II {

    public static void main(String[] args) {

    }

    /**
     * 贪心
     * <p>
     * 将股票利润分成一段一段的序列 我们只取其中大于0的序列 将他们加起来 这样就能得到最大的利润 从局部推到整体
     *
     * @param prices
     * @return
     */
    public int maxProfit(int[] prices) {
        int result = 0;
        for (int i = 1; i < prices.length; i++) {
            result += Math.max(prices[i] - prices[i - 1], 0);
        }
        return result;
    }
}

File: golang-devildyw/middle/L122买卖股票的最佳时机II.go
package main

import "math"

func main() {

}

func maxProfit(prices []int) int {
	// 可以多次买卖股票 但同一时刻最多只能持有一股（也可以不持有）
	// 在这种情况下想要获得最大利润 可以把问题拆解为递推状态转移 最大利润需要由前面的最大利润得到 也就是转化为了最小子问题最优解
	// 在一天里 你可以选择买出股票 或者买入股票 也可以保持原样

	//根据上面的信息我们可以确定状态了
	// dp[i][0] 为第i天不持有股票的利润
	// dp[i][1] 为第i天持有股票的利润
	var (
		dp [][2]int = make([][2]int, len(prices))
	)
	dp[0][0] = 0
	dp[0][1] = -prices[0]

	for i := 1; i < len(prices); i++ {
		dp[i][0] = int(math.Max(float64(dp[i-1][0]), float64(prices[i]+dp[i-1][1])))
		dp[i][1] = int(math.Max(float64(dp[i-1][1]), float64(dp[i-1][0]-prices[i])))
	}

	return dp[len(prices)-1][0]
}
